File: src/stores/gateway.ts
/**
 * Gateway connection state store.
 *
 * Tracks the WebSocket connection lifecycle, URL, version info,
 * and reconnection state.
 */

import { create } from "zustand";
import type { ConnectionState } from "../types/gateway";
import {
  connectGateway,
  disconnectGateway,
  getToken,
} from "../services/tauri-commands";

interface GatewayStore {
  /** Current WebSocket connection state. */
  connectionState: ConnectionState;
  /** Gateway URL (e.g., "ws://localhost:18789"). */
  url: string;
  /** Gateway version string, if known. */
  version: string | null;
  /** Gateway uptime in seconds, if known. */
  uptime: number | null;
  /** Current reconnection attempt number. */
  reconnectAttempt: number;
  /** Last error message. */
  lastError: string | null;

  // Actions
  /** Initiate a connection to the Gateway. */
  connect: (url: string, token: string) => Promise<void>;
  /** Disconnect from the Gateway. */
  disconnect: () => Promise<void>;
  /** Auto-connect using stored token. */
  autoConnect: (url?: string) => Promise<void>;
  /** Update the connection state. */
  setConnectionState: (state: ConnectionState) => void;
  /** Update the reconnection attempt number. */
  setReconnectAttempt: (attempt: number) => void;
  /** Update the last error. */
  setLastError: (error: string | null) => void;
  /** Update gateway info (version, uptime). */
  setGatewayInfo: (version: string, uptime: number) => void;
}

export const useGatewayStore = create<GatewayStore>((set, get) => ({
  connectionState: "disconnected",
  url: "ws://localhost:18789",
  version: null,
  uptime: null,
  reconnectAttempt: 0,
  lastError: null,

  connect: async (url: string, token: string) => {
    set({ connectionState: "connecting", url, lastError: null });
    try {
      await connectGateway(url, token);
    } catch (err) {
      set({
        connectionState: "disconnected",
        lastError: err instanceof Error ? err.message : String(err),
      });
    }
  },

  disconnect: async () => {
    try {
      await disconnectGateway();
    } catch (err) {
      console.error("Disconnect error:", err);
    }
    set({ connectionState: "disconnected", reconnectAttempt: 0 });
  },

  autoConnect: async (url?: string) => {
    const gatewayUrl = url || get().url;
    try {
      const token = await getToken();
      if (token) {
        await get().connect(gatewayUrl, token);
      }
    } catch (err) {
      console.error("Auto-connect failed:", err);
    }
  },

  setConnectionState: (state: ConnectionState) =>
    set({ connectionState: state }),

  setReconnectAttempt: (attempt: number) =>
    set({ reconnectAttempt: attempt }),

  setLastError: (error: string | null) => set({ lastError: error }),

  setGatewayInfo: (version: string, uptime: number) =>
    set({ version, uptime }),
}));

File: src-tauri/src/commands/gateway.rs
//! Gateway connection and RPC commands.
//!
//! Provides Tauri commands for establishing/closing the WebSocket
//! connection and making RPC calls to the Gateway.

use crate::gateway::client;
use crate::state::{AppState, ConnectionState, WsCommand};
use serde_json::Value;
use std::sync::Arc;
use tauri::State;

/// Establish a WebSocket connection to the Gateway.
///
/// Spawns a background connection manager task that handles
/// authentication, message routing, and reconnection.
#[tauri::command]
pub fn connect_gateway(
    app: tauri::AppHandle,
    state: State<'_, Arc<AppState>>,
    url: String,
    token: String,
) -> Result<(), String> {
    let current = state.connection_state.read().clone();
    if current == ConnectionState::Connected
        || current == ConnectionState::Connecting
        || current == ConnectionState::Reconnecting
    {
        // Disconnect existing connection first if reconnecting
        if current == ConnectionState::Reconnecting {
            let ws_tx = state.ws_command_tx.read();
            if let Some(tx) = ws_tx.as_ref() {
                let _ = tx.send(WsCommand::Disconnect);
            }
            drop(ws_tx);
            // Brief delay to let the old loop clean up
            std::thread::sleep(std::time::Duration::from_millis(100));
        } else {
            return Err("Already connected or connecting".to_string());
        }
    }

    // Store credentials in state
    {
        let mut gw_url = state.gateway_url.write();
        *gw_url = Some(url.clone());
        let mut auth = state.auth_token.write();
        *auth = Some(token.clone());
    }

    client::start_connection(app, Arc::clone(&state), url, token);
    Ok(())
}

/// Disconnect from the Gateway WebSocket.
///
/// Sends a close frame and stops the connection manager.
/// Does not attempt reconnection after a manual disconnect.
#[tauri::command]
pub fn disconnect_gateway(state: State<'_, Arc<AppState>>) -> Result<(), String> {
    let ws_tx = state.ws_command_tx.read();
    if let Some(tx) = ws_tx.as_ref() {
        tx.send(WsCommand::Disconnect)
            .map_err(|_| "Failed to send disconnect command".to_string())?;
    }
    Ok(())
}

/// Send an RPC call to the Gateway over the WebSocket.
///
/// Blocks until the response is received or a 30-second timeout occurs.
///
/// # Arguments
///
/// * `method` - The RPC method name (e.g., "health", "cron.list")
/// * `params` - Method parameters as a JSON value
#[tauri::command]
pub async fn gateway_rpc(
    state: State<'_, Arc<AppState>>,
    method: String,
    params: Value,
) -> Result<Value, String> {
    client::send_rpc(&state, method, params).await
}

/// Get the current Gateway connection state.
#[tauri::command]
pub fn get_connection_state(state: State<'_, Arc<AppState>>) -> String {
    let conn = state.connection_state.read();
    match *conn {
        ConnectionState::Disconnected => "disconnected".to_string(),
        ConnectionState::Connecting => "connecting".to_string(),
        ConnectionState::Connected => "connected".to_string(),
        ConnectionState::Reconnecting => "reconnecting".to_string(),
    }
}
